File: Python3/1466. Reorder Routes to Make All Paths Lead to the City Zero.py
class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        # add pseudo path to make each edge bidirected
        # the pseudo paths don't impact on the counting, mark as 0
        # only increment the answer when encounter an actual reversed edge, mark as 1
        adj = collections.defaultdict(list)
        for u, v in connections:
            adj[u].append((v, 1))
            adj[v].append((u, 0))

        q = collections.deque([0])
        ans = 0
        visited = set([0])

        while q:
            node = q.popleft()
            for child, sign in adj[node]:
                if child not in visited:
                    ans += sign
                    q.append(child)
                    visited.add(child)

        return ans

File: Java/1466. Reorder Routes to Make All Paths Lead to the City Zero.java
class Solution {
    public int minReorder(int n, int[][] connections) {
        Map<Integer, List<int[]>> adj = new HashMap();
        for (int[] c: connections) {
            adj.computeIfAbsent(c[0], value -> new ArrayList()).add(new int[]{c[1], 1});
            adj.computeIfAbsent(c[1], value -> new ArrayList()).add(new int[]{c[0], 0});
        }

        Queue<Integer> q = new LinkedList();
        q.add(0);
        Set<Integer> visited = new HashSet();
        visited.add(0);
        int ans = 0;

        while (!q.isEmpty()) {
            int node = q.poll();
            if (!adj.containsKey(node)) {
                continue;
            }

            for (int[] p: adj.get(node)) {
                if (!visited.contains(p[0])) {
                    visited.add(p[0]);
                    q.offer(p[0]);
                    ans += p[1];
                }
            }
        }

        return ans;
    }
}